### ISSUE TITLE
2021.10-003 rockpi4c image doesn't boot - GPT  signature is wrong

### ISSUE THREAD

**Issue by danboid** [Tow-Boot/Tow-Boot#112]
I have tried booting the radxa-RockPi4C-2021.10-003 spi.installer.img off 2 different uSD cards on my rockpi4c. I don't see any errors when I try to boot it nor do I get any picture via micro HDMI. Over the serial port I get this:

```
DDR Version 1.20 20190314
In
channel 0
CS = 0
MR0=0x18
MR4=0x1
MR5=0x1
MR8=0x8
MR12=0x72
MR14=0x72
MR18=0x0
MR19=0x0
MR24=0x8
MR25=0x0
CS = 1
MR0=0x18
MR4=0x1
MR5=0x1
MR8=0x8
MR12=0x72
MR14=0x72
MR18=0x0
MR19=0x0
MR24=0x8
MR25=0x0
channel 1
CS = 0
MR0=0x18
MR4=0x1
MR5=0x1
MR8=0x8
MR12=0x72
MR14=0x72
MR18=0x0
MR19=0x0
MR24=0x8
MR25=0x0
CS = 1
MR0=0x18
MR4=0x1
MR5=0x1
MR8=0x8
MR12=0x72
MR14=0x72
MR18=0x0
MR19=0x0
MR24=0x8
MR25=0x0
channel 0 training pass!
channel 1 training pass!
change freq to 400MHz 0,1
channel 0
CS = 0
MR0=0x18
MR4=0x1
MR5=0x1
MR8=0x8
MR12=0x72
MR14=0x72
MR18=0x0
MR19=0x0
MR24=0x8
MR25=0x0
CS = 1
MR0=0x18
MR4=0x1
MR5=0x1
MR8=0x8
MR12=0x72
MR14=0x72
MR18=0x0
MR19=0x0
MR24=0x8
MR25=0x0
channel 1
CS = 0
MR0=0x18
MR4=0x1
MR5=0x1
MR8=0x8
MR12=0x72
MR14=0x72
MR18=0x0
MR19=0x0
MR24=0x8
MR25=0x0
CS = 1
MR0=0x18
MR4=0x1
MR5=0x1
MR8=0x8
MR12=0x72
MR14=0x72
MR18=0x0
MR19=0x0
MR24=0x8
MR25=0x0
channel 0 training pass!
channel 1 training pass!
change freq to 800MHz 1,0
Channel 0: LPDDR4,800MHz
Bus Width=32 Col=10 Bank=8 Row=15/15 CS=2 Die Bus-Width=16 Size=2048MB
Channel 1: LPDDR4,800MHz
Bus Width=32 Col=10 Bank=8 Row=15/15 CS=2 Die Bus-Width=16 Size=2048MB
256B stride
ch 0 ddrconfig = 0x101, ddrsize = 0x2020
ch 1 ddrconfig = 0x101, ddrsize = 0x2020
pmugrf_os_reg[2] = 0x3AA1FAA1, stride = 0xD
OUT
Boot1: 2018-06-26, version: 1.14
CPUId = 0x0
ChipType = 0x10, 221
Spi_ChipId = b4016
SpiBootInit:0
mmc0:cmd5,32
SdmmcInit=0 0
BootCapSize=0
UserCapSize=30436MB
FwPartOffset=2000 , 0
StorageInit ok = 151568
SecureMode = 0
SecureInit ret = 0, SecureMode = 0
GPT  signature is wrong
LoadTrust Addr:0x1800
LoadTrust Addr:0x1c00
LoadTrust Addr:0x2000
LoadTrust Addr:0x2400
LoadTrust Addr:0x2800
LoadTrust Addr:0x2c00
LoadTrust Addr:0x3000
LoadTrust Addr:0x3400
Addr:0x1800 No find trust.img!
LoadTrustBL error:-3
SecureMode = 0
SecureInit ret = 0, SecureMode = 0
```

**Comment by danboid** [comment]
Does this image boot for you @CRTified ? I flashed the uSD cards using balena etcher so it verifies them by default.

**Comment by CRTified** [comment]
I did not have time to test the released image so far (The 4C runs my home automation and DNS, so it's hard to find a good time where it does not cause problems).

Just as a sidenote: On the matrix channel, there was another report about the image not booting.

The latest version I tested before the release was [the one in this PR](https://github.com/Tow-Boot/Tow-Boot/pull/56#issuecomment-999205954). But that one was self-built. Would be nice to know whether there were any problems.

As an alternative, you might want to try the version from the `development`-branch (where a [lot of refactoring](https://github.com/Tow-Boot/Tow-Boot/pull/78) was done). I know of at least one user of an RockPi4A (OP1) that successfully uses the version on rev `1297c1bbb6ba882d5aa79f935fe1e55c243f639d`, and the device trees of the 4A, 4B and 4C only differ marginally.

Sorry for the inconvenience. I'll try to see whether I have time to test it in the next few days.

**Comment by danboid** [comment]
Sounds like it might be best that I try building a dev image first, but how? I've looked at the README and the docs dir but I don't see any build instructions.

How do I build an image for the rockpi4c?

**Comment by CRTified** [comment]
If you are under NixOS or have `nix` available on your distribution (which is AFAICT the only supported build method right now), it's a mere `nix-build -A radxa-RockPi4C`.

If I remember correctly, @samueldr was thinking about some other build options like a nix-based docker image, but I don't know about the progress there.

**Comment by danboid** [comment]
I've never used nix or NixOS.

Can I use nix under Ubuntu amd64 to build towboot for my rockpi4c or will I need to build towboot on my rockipi4c (or cross compile, but I'm not going that route)?

**Comment by samueldr** [comment]
> ```
> DDR Version 1.20 20190314
> In
> channel 0
> [...]
> ```

This looks like the proprietary Rockchip RAM training blobs. I assume the Tow-Boot present on SD card is not started.

Note that if you have a U-Boot starting when no SD card is inserted, it will need to be bypassed *in some way* (usually device/vendor-dependent) to start the one on the SD card.

**Comment by CRTified** [comment]
For bypassing, it *should* be possible by killing the SPI clock. In that case, have a look [at the GPIO pinout](https://wiki.radxa.com/Rockpi4/hardware/gpio) and [the SPI documentation](https://wiki.radxa.com/Rockpi4/hardware/spi_flash).
Should be sufficient to tie pin `23` to GND (using a small resistor would be perfect, but shouldn't be necessary). You *will* need to remove that connection when attempting to communicate with the SPI flash.

Alternatively, you could try a different kernel that supports the SPI flash from user/kernelspace, but that'll likely require a patched kernel and building it on your own.


For building: You can use amd64 without any problems, and `nix` is available there (At this point, it's sufficient to look at its use for Tow-Boot as a glorified `Makefile` that tracks dependencies). 

In case you trust me enough, you could also use the attached file (zipped as github won't allow me to upload `*.img` directly). I've built it on the previously mentioned revision of the `development`-branch (which is known-working for the 4A/B).

[spi.installer.zip](https://github.com/Tow-Boot/Tow-Boot/files/8163101/spi.installer.zip)

**Comment by danboid** [comment]
I get the same result using the image in that zip.

I have tried putting a jumper over pins 23+25 but it doesn't seem to boot at all then.

Should I expect to see the towboot install menu via HDMI or is it something that requires a UART cable to view?

**Comment by CRTified** [comment]
It should be visible on both UART and HDMI, although I've installed it over UART.

**Comment by gattohub** [comment]
Hello. I have tried to install radxa-RockPi4C-2021.10-003 spi.installer.img on a rockpi4b with Armbian.
I had followed this link to get booting from NVME without SD card with Armbian:  

https://forum.armbian.com/topic/16195-booting-rock-pi-4abc-with-mainline-u-boot-in-spi-nvme-and-armbian-v2011x/

And I have been booting Armbian from NVME without sd card with no problem. But now I would like to do the same with Manjaro, and I can't get it.
Now Armbian stopped supporting these boards.  That's why I've been trying to install Tow-boot.
But failed to boot from sd with radxa-RockPi4C-2021.10-003 spi.installer.img
And from what I've been reading, it should work with rockpi4B.
I haven't tried bridging in gpio, but apparently it didn't work either from what you've been talking about here.
Google translator Has been used

**Comment by danboid** [comment]
@gattohub I'm sure @CRTified will correct me if I'm wrong but I don't think towboot supports the 4b unless it has a SPI chip.

I get the impression there were different revisions of the 4b and some don't have the SPI.

I'm going to be trying to build towboot dev later, I will update this thread after I've tried it.

**Comment by CRTified** [comment]
> I get the impression there were different revisions of the 4b and some don't have the SPI.

That is correct. [According to the official docs](https://wiki.radxa.com/Rockpi4/hardware/spi_flash), the SPI was not populated up to revision 1.3. But it's easy to solder in on your own (`GD25Q127CSIG` is relatively easy to source, the other chips mentioned there not so much).

> I'm sure @CRTified will correct me if I'm wrong but I don't think towboot supports the 4b unless it has a SPI chip.

Tow-Boot should be able to be used in the place of a "classical" u-boot image on the eMMC or µSD-Card, but I did not test this.

**Comment by samueldr** [comment]
Yes, the `shared.disk-image.img` is just that. It's pre-seeded with a GPT that covers the platform firmware partition with a protective partition for enhanced EBBR conformity.

**Comment by danboid** [comment]
I have builtt the `development` branch using the command @CRTified mentined above but its still not booting. I get no output via HDMI and this on the console, which seems to be the same as with the latest release build:

```
DDR Version 1.20 20190314
In
channel 0
CS = 0
MR0=0x18
MR4=0x1
MR5=0x1
MR8=0x8
MR12=0x72
MR14=0x72
MR18=0x0
MR19=0x0
MR24=0x8
MR25=0x0
CS = 1
MR0=0x18
MR4=0x1
MR5=0x1
MR8=0x8
MR12=0x72
MR14=0x72
MR18=0x0
MR19=0x0
MR24=0x8
MR25=0x0
channel 1
CS = 0
MR0=0x18
MR4=0x1
MR5=0x1
MR8=0x8
MR12=0x72
MR14=0x72
MR18=0x0
MR19=0x0
MR24=0x8
MR25=0x0
CS = 1
MR0=0x18
MR4=0x1
MR5=0x1
MR8=0x8
MR12=0x72
MR14=0x72
MR18=0x0
MR19=0x0
MR24=0x8
MR25=0x0
channel 0 training pass!
channel 1 training pass!
change freq to 400MHz 0,1
channel 0
CS = 0
MR0=0x18
MR4=0x1
MR5=0x1
MR8=0x8
MR12=0x72
MR14=0x72
MR18=0x0
MR19=0x0
MR24=0x8
MR25=0x0
CS = 1
MR0=0x18
MR4=0x1
MR5=0x1
MR8=0x8
MR12=0x72
MR14=0x72
MR18=0x0
MR19=0x0
MR24=0x8
MR25=0x0
channel 1
CS = 0
MR0=0x18
MR4=0x1
MR5=0x1
MR8=0x8
MR12=0x72
MR14=0x72
MR18=0x0
MR19=0x0
MR24=0x8
MR25=0x0
CS = 1
MR0=0x18
MR4=0x1
MR5=0x1
MR8=0x8
MR12=0x72
MR14=0x72
MR18=0x0
MR19=0x0
MR24=0x8
MR25=0x0
channel 0 training pass!
channel 1 training pass!
change freq to 800MHz 1,0
Channel 0: LPDDR4,800MHz
Bus Width=32 Col=10 Bank=8 Row=15/15 CS=2 Die Bus-Width=16 Size=2048MB
Channel 1: LPDDR4,800MHz
Bus Width=32 Col=10 Bank=8 Row=15/15 CS=2 Die Bus-Width=16 Size=2048MB
256B stride
ch 0 ddrconfig = 0x101, ddrsize = 0x2020
ch 1 ddrconfig = 0x101, ddrsize = 0x2020
pmugrf_os_reg[2] = 0x3AA1FAA1, stride = 0xD
OUT
Boot1: 2018-06-26, version: 1.14
CPUId = 0x0
ChipType = 0x10, 221
Spi_ChipId = b4016
SpiBootInit:0
mmc0:cmd5,32
SdmmcInit=0 0
BootCapSize=0
UserCapSize=30436MB
FwPartOffset=2000 , 0
StorageInit ok = 151285
SecureMode = 0
SecureInit ret = 0, SecureMode = 0
GPT  signature is wrong
LoadTrust Addr:0x1800
LoadTrust Addr:0x1c00
LoadTrust Addr:0x2000
LoadTrust Addr:0x2400
LoadTrust Addr:0x2800
LoadTrust Addr:0x2c00
LoadTrust Addr:0x3000
LoadTrust Addr:0x3400
Addr:0x1800 No find trust.img!
LoadTrustBL error:-3
SecureMode = 0
SecureInit ret = 0, SecureMode = 0
```

**Comment by danboid** [comment]
I think I'll have to install Tow-Boot via the rpi4 mask rom case 2 procedure like this:

https://wiki.radxa.com/Rockpi4/dev/spi-install

I have got my rockpi4 to boot in maskrom mode but I've not been able to use rkdeveloptool to write it:

```
$ rkdeveloptool ld
DevNo=1	Vid=0x2207,Pid=0x330c,LocationID=101	Maskrom
# rkdeveloptool wl 0 /home/dan/spi.installer.img
Write LBA failed, err=13, can't open file: /home/dan/spi.installer.img
```

I installed rkdeveloptool via a snap under Ubuntu. I wonder if I build it from source or use a PPA to install rkdeveloptool if it'll work?

**Comment by danboid** [comment]
Looks like I'll have to use the Windows Rockchip tool as I'm not having any luck building it from source under Ubuntu 20.04:

```
~/src/rkdeveloptool$ make
make[1]: Entering directory '/home/dan/src/rkdeveloptool'
g++ -DHAVE_CONFIG_H -I. -I./cfg  -Wall -Werror -Wextra -Wreturn-type -fno-strict-aliasing -D_FILE_OFFSET_BITS=64 -D_LARGE_FILE -I/usr/include/libusb-1.0   -g -O2 -MT main.o -MD -MP -MF .deps/main.Tpo -c -o main.o main.cpp
main.cpp: In function ‘bool _Z9mergeBootv.part.0()’:
main.cpp:1493:36: error: ‘%s’ directive output may be truncated writing up to 557 bytes into a region of size 5 [-Werror=format-truncation=]
 1493 |  snprintf(buffer, sizeof(buffer), "%s", chip);
      |                                    ^~
......
 1534 |   chipType = convertChipType(chip + 2);
      |              ~~~~~~~~~~~~~~~~~~~~~~~~~
In file included from /usr/include/stdio.h:867,
                 from DefineHeader.h:3,
                 from main.cpp:11:
/usr/include/x86_64-linux-gnu/bits/stdio2.h:67:35: note: ‘__builtin_snprintf’ output between 1 and 558 bytes into a destination of size 5
   67 |   return __builtin___snprintf_chk (__s, __n, __USE_FORTIFY_LEVEL - 1,
      |          ~~~~~~~~~~~~~~~~~~~~~~~~~^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
   68 |        __bos (__s), __fmt, __va_arg_pack ());
      |        ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
cc1plus: all warnings being treated as errors
make[1]: *** [Makefile:489: main.o] Error 1
make[1]: Leaving directory '/home/dan/src/rkdeveloptool'
make: *** [Makefile:509: all-recursive] Error 1
```

**Comment by gattohub** [comment]
well, i can confirm that with the armbiam uboot you get from the link i used above, you can boot from nvme with manjaro without sd.

**Comment by danboid** [comment]
@gattohub 

Good news! It doesn't seem that Armbian ever released any install images for the 4c but I may be able to boot one of the archived 4b images or maybe build a 4c image with the build script and do what you've done?

**Comment by CRTified** [comment]
@gattohub Just as a sidenote: The development branch also has an image for the 4A/B (they share their dtb).

**Comment by danboid** [comment]
There is a pull request to fix building rkdeveloptool here:

https://github.com/rockchip-linux/rkdeveloptool/pull/63

I still cannot flash towboot using the latest rkdeveloptool built from source:

```
./rkdeveloptool wl 0 /home/dan/spi.installer.img
Write LBA failed!
```

I have tried removing the jumper from the pins after getting it to boot in maskrom mode but I still get the Write LBA failed error.

I will try using the Windows tool next.

**Comment by danboid** [comment]
I tried RKDevTool under Windows 10 and I got no further than under Linux. It can see my board is connected in maskrom mode but nothing else worked such as trying to erase to flash, download the bootloader or install the firmware.

**Comment by danboid** [comment]
I can confirm that the instructions to install the armbian u-boot that @gattohub linked to above also work on the 4c and I can now boot directly off my Samsung EVO 960 SSD. I used Armbian_21.08.1_Rockpi-4b_hirsute_edge_5.13.12_xfce_desktop.img to install the Armbian u-boot.

@gattohub - did you have to / could you successfully run:

```
echo spi1.0 > /sys/bus/spi/drivers/spi-nor/bind
```

I got an error about no such device existing after running that command but the rest of the instructions worked anyway. I will probably update the Rockpi4 wiki so I'd like to know if this command is required on the 4b?

**Comment by samueldr** [comment]
> ```
> ./rkdeveloptool wl 0
> /home/dan/spi.installer.img Write LBA failed!
> ```

This would have written the *installer image*, I believe to eMMC. I'm not sure `rkdeveloptool` has facilities to write to SPI.

**Comment by joaobertacchi** [comment]
Same problem here. 2021.10-004 rockpi4c spi.installer.img also does not boot.
